File: AdaptationsToFrameworks/Blazor/Package/GUI_Components/Controls/Validation/PreMadeValidations/PasswordInputtedValueValidation.cs
using YamatoDaiwa.Frontend.GUI_Components.Controls.Validation.PreMadeRules.Strings;

namespace YamatoDaiwa.Frontend.GUI_Components.Controls.Validation.PreMadeValidations;


public class PasswordInputtedValueValidation(
  IConvertible minimalCharactersCount,
  IConvertible maximalCharactersCount,
  bool? inputRequiredFlag = null,
  Func<bool>? isInputRequired = null,
  PasswordInputtedValueValidation.ILocalization? localization = null,
  IEnumerable<char>? allowedNonWordCharacters = null,
  System.Text.RegularExpressions.Regex? regularExpression = null,
  InputtedValueValidation.IRule[]? contextDependentRules = null,
  InputtedValueValidation.IAsynchronousRule[]? asynchronousRules = null,
  Action<string>? asynchronousValidationFailureLogger = null
) : 
  
  InputtedValueValidation(
    
    isValueOfSupportedType: (object? rawValue) => rawValue is string,
    
    hasValueBeenOmitted: (object? rawValue) => !String.IsNullOrEmpty((string?) rawValue),
    
    inputRequiredFlag,
    
    isInputRequired,
    
    requiredInputIsMissingValidationErrorMessage:
        (localization ?? PasswordInputtedValueValidation.Localization).RequiredInputIsMissingValidationErrorMessage,
    
    staticRules: 
    
        [
        
          new AllowedCharactersInputtedValueValidationRule
          {
            AllowedCharactersGroups = new AllowedCharactersInputtedValueValidationRule.AllowedCharactersSpecification()
            {
              latinUppercase = true,
              latinLowercase = true,
              digits = true,
              other = allowedNonWordCharacters ?? []
            },
            ErrorMessageBuilder = 
                (localization ?? PasswordInputtedValueValidation.Localization).
                    AllowedCharactersValidationErrorMessageBuilder,
          },
          
          new MinimalCharactersCountInputtedValueValidationRule
          {
            MinimalCharactersCount = minimalCharactersCount,
            ErrorMessageBuilder =
                (localization ?? PasswordInputtedValueValidation.Localization).
                    MinimalCharactersCountValidationErrorMessageBuilder,
            MustFinishValidationIfValueIsInvalid = true
          },

          new MaximalCharactersCountInputtedValueValidationRule
          {
            MaximalCharactersCount = maximalCharactersCount,
            ErrorMessageBuilder =
                (localization ?? PasswordInputtedValueValidation.Localization).
                    MaximalCharactersCountValidationErrorMessageBuilder,
          }
          
        ],
    
    contextDependentRules,
    
    asynchronousRules,
    
    asynchronousValidationFailureLogger
    
  )

{
  
  public readonly IConvertible MINIMAL_CHARACTERS_COUNT = minimalCharactersCount;
  public readonly IConvertible MAXIMAL_CHARACTERS_COUNT = maximalCharactersCount;
  
  
  /* ━━━ Localization ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ */
  public new interface ILocalization: InputtedValueValidation.ILocalization
  {
    
    public Func<AllowedCharactersInputtedValueValidationRule.ILocalization.ErrorMessage.TemplateVariables, string> 
        AllowedCharactersValidationErrorMessageBuilder { get; }
    
    public Func<MinimalCharactersCountInputtedValueValidationRule.ILocalization.ErrorMessage.TemplateVariables, string> 
        MinimalCharactersCountValidationErrorMessageBuilder { get; }
    
    public Func<MaximalCharactersCountInputtedValueValidationRule.ILocalization.ErrorMessage.TemplateVariables, string> 
        MaximalCharactersCountValidationErrorMessageBuilder { get; }
    

  }
  
  public new static ILocalization Localization = new PasswordInputtedValueValidationEnglishLocalization();
  
}

File: CoreLibrary/Package/GUI_Components/Controls/_Validation/PreMadeValidations/Strings/Password/PasswordInputtedValueValidation.ts
import InputtedValueValidation from "../../../InputtedValueValidation";

import MinimalCharactersCountInputtedValueValidationRule from
    "../../../PreMadeRules/Strings/MinimalCharactersCountInputtedValueValidationRule";
import MaximalCharactersCountInputtedValueValidationRule from
    "../../../PreMadeRules/Strings/MaximalCharactersCountInputtedValueValidationRule";
import AllowedCharactersInputtedValueValidationRule from
    "../../../PreMadeRules/Strings/AllowedCharactersInputtedValueValidationRule";

import { isString } from "@yamato-daiwa/es-extensions";


abstract class PasswordInputtedValueValidation<EmptyValue extends PasswordInputtedValueValidation.SupportedEmptyValues = string>
    extends InputtedValueValidation<string, EmptyValue>
/* eslint-disable-next-line @stylistic/brace-style -- Allow Allman style for square areas principle. */
{

  public static localization: PasswordInputtedValueValidation.Localization;

  public readonly MINIMAL_CHARACTERS_COUNT: number;
  public readonly MAXIMAL_CHARACTERS_COUNT: number;


  protected constructor(
    {
      minimalCharactersCount,
      maximalCharactersCount,
      isInputRequired,
      localization,
      allowedNonWordCharacters,
      contextDependentRules,
      asynchronousRules,
      asynchronousValidationsCallback,
      ...compoundParameter
    }: PasswordInputtedValueValidation.ConstructorParameter<EmptyValue>
  ) {

    super({

      isValueOfSupportedType:
          "isValueOfSupportedType" in compoundParameter ?
              compoundParameter.isValueOfSupportedType :
              isString,

      hasValueBeenOmitted:
          "hasValueBeenOmitted" in compoundParameter ?
              compoundParameter.hasValueBeenOmitted :
              /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
              * It is unlikely that overloading functionality in TypeScript allows expressing: "When `omittedValueChecker`
              *   is specified, the return type will the `PasswordInputtedValueValidation<EmptyValue>`, otherwise
              *   it will be `PasswordInputtedValueValidation<string>`", so maybe have no choice except casting. */
              (
                (supportedButPossiblyEmptyValue: string): supportedButPossiblyEmptyValue is string =>
                    supportedButPossiblyEmptyValue.length === 0
              ) as InputtedValueValidation.OmittedValueChecker<string, EmptyValue>,

      isInputRequired,

      requiredInputIsMissingValidationErrorMessage:
          (localization ?? PasswordInputtedValueValidation.localization).requiredInputIsMissingValidationErrorMessage,

      staticRules: [

        new AllowedCharactersInputtedValueValidationRule({
          allowedCharacters: {
            latinLowercase: true,
            latinUppercase: true,
            digits: true,
            other: allowedNonWordCharacters
          },
          errorMessageBuilder:
              (localization ?? PasswordInputtedValueValidation.localization).allowedCharactersValidationErrorMessageBuilder
        }),

        new MinimalCharactersCountInputtedValueValidationRule({
          minimalCharactersCount,
          errorMessageBuilder:
              (localization ?? PasswordInputtedValueValidation.localization).
                  minimalCharactersCountValidationErrorMessageBuilder,
          mustFinishValidationIfValueIsInvalid: true
        }),

        new MaximalCharactersCountInputtedValueValidationRule({
          maximalCharactersCount,
          errorMessageBuilder:
              (localization ?? PasswordInputtedValueValidation.localization).
                  maximalCharactersCountValidationErrorMessageBuilder
        })

      ],

      contextDependentRules,

      asynchronousRules,

      asynchronousValidationsCallback

    });

    this.MINIMAL_CHARACTERS_COUNT = minimalCharactersCount;
    this.MAXIMAL_CHARACTERS_COUNT = maximalCharactersCount;

  }

}


namespace PasswordInputtedValueValidation {

  export type SupportedEmptyValues = string | null | undefined;

  export type ConstructorParameter<EmptyValueType extends SupportedEmptyValues> =
      ConstructorParameter.EmptyValueIsString |
      ConstructorParameter.EmptyValueIsNonString<EmptyValueType>;

  export namespace ConstructorParameter {

    export type EmptyValueIsString =

        Readonly<{
          isInputRequired: boolean;
          minimalCharactersCount: number;
          maximalCharactersCount: number;
          allowedNonWordCharacters: ReadonlyArray<string>;
          localization?: Localization;
        }> &

        Pick<
          InputtedValueValidation.ConstructorCompoundParameter<string>,
          "contextDependentRules" |
          "asynchronousRules" |
          "asynchronousValidationsCallback"
        >;

    export type EmptyValueIsNonString<EmptyValueType extends SupportedEmptyValues> =

        EmptyValueIsString &

        Readonly<{
          isValueOfSupportedType: InputtedValueValidation.SupportedValueChecker<string, EmptyValueType>;
          hasValueBeenOmitted: InputtedValueValidation.OmittedValueChecker<string, EmptyValueType>;
        }>;

  }


  export type Localization =
      InputtedValueValidation.Localization &
      Readonly<{
        allowedCharactersValidationErrorMessageBuilder:
            AllowedCharactersInputtedValueValidationRule.ErrorMessage.Builder;
        minimalCharactersCountValidationErrorMessageBuilder:
            MinimalCharactersCountInputtedValueValidationRule.ErrorMessage.Builder;
        maximalCharactersCountValidationErrorMessageBuilder:
            MaximalCharactersCountInputtedValueValidationRule.ErrorMessage.Builder;
      }>;

}


export default PasswordInputtedValueValidation;
